### ISSUE TITLE
allow overriding workdir for container jobs

### ISSUE THREAD

**Issue by igagis** [actions/runner#878]
**Describe the enhancement**
Allow overriding container's workdir.
Currently, even if passing `--workdir=/somedir` to the `container: options:` it does not override the first `--workdir` option passed to `docker create` by the runner.
It would be nice to be able to override the `workdir` somehow, so that on self-hosted runners, after the job is done the checked out sources would not remain the in `actions-runner/_work` directory. In container jobs it makes no sense to mount the external directory, as all the job steps are executed within the container anyway. If work dir is set to some in-container directory, then after the job is done the container is discarded and all the build artifacts are discarded as well, no separate cleanup is needed.

**Code Snippet**
````
container:
  image: debian:buster
  options: --workdir=/home/build/myrepo
````
This example currently will result in that `--workdir` is passed twice to the `docker create`. The first one is the default on which is `-workdir /__w/myrepo/myrepo` and the second is the one indicated in `options:`, but it looks like `docker create` takes the first one and ignores the second one. So, there is no possibility to override the `workdir`.
Need to exclude the default `--workdir` option in case `options:` contains custom `--workdir` option.

**Comment by crazy-matt** [comment]
Same thing observed on my side. The documentation is referring to the [docker create options](https://docs.docker.com/engine/reference/commandline/create/#options). 
I tried to override the HOME env var with:
```
container:
      image: ${{ matrix.image }}
      credentials:
        username: ${{ github.repository_owner }}
        password: ${{ secrets.PAT }}
      options: --user cicd -e "HOME=/home/cicd"
```

But I still get these env vars:
```
HOME=/github/home
GITHUB_ACTIONS=true
CI=true
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
BASE_IMAGE=
LANG=en_GB.UTF-8
LC_ALL=C
LANGUAGE=en_GB:en
```

**Comment by yurirocha15** [comment]
I am having a similar issue. I need to place my code inside a specific directory in order to build it (which is the default workdir in my container).
In my opinion, the runner should only set the workdir/HOME if they aren't set (or provide a way to change the default values).

**Comment by lidongze0629** [comment]
Does anyone have some idea to change the `workdir` in container runner ? Thanks.

**Comment by MarcelWilson** [comment]
Yes please.  While it can be convenient to have workdir default to `/github/workspace`, the fact that there isn't an override forces editing existing Dockerfiles that already establish the WORKDIR.

I found it cute that the documentation has this little note in it; without giving any explanation as to _why_ this is being done.  

>  It's recommended to not use the `WORKDIR` instruction in your `Dockerfile`

**Comment by ethomson** [comment]
Yeah, I think that this makes sense.  Today you can use a workaround of `docker run`, since `container` jobs are just a veneer on top of that, but we'll need to give this more thought if we do something different for k8s runners vs static runners.  We'll put this on the backlog.

**Comment by joshjohanning** [comment]
Can confirm, we were having a hard time with being able to override the `--workdir` parameter, so therefore our `actions/checkout` action was trying to clone in a `/_ww/*` folder that we don't have permissions to write to in the container. 

We can't modify the `actions/checkout` path because it is a _relative_ path to the working directory (so we can't override to `/usr/*` for example we can only pick a different subfolder under `/_ww/*`.

It would be nice to override that parameter, or provide guidance on how we can default the checkout to a different directory on the container.

**Comment by joshjohanning** [comment]
We resolved a problem of ours with insufficient permissions in the volume mapped to `/__w` by granting sufficient permissions (`chmod 777 /_work`) on the *host*. This was only affecting our self-hosted runners, not our cloud runners. 

Relevant errors:

>   /usr/bin/git init /__w/container-job-test/container-job-test
>  /__w/container-job-test/container-job-test/.git: Permission denied
>  Error: The process '/usr/bin/git' failed with exit code 1

and

> Deleting the contents of '/__w/container-job-test/container-job-test'
> Error: Command failed: rm -rf "/__w/container-job-test/container-job-test/.git"
> rm: cannot remove '/__w/container-job-test/container-job-test/.git': Permission denied

**Comment by vwcruzeiro** [comment]
Not being able to reset `--workdir` is a big problem when there are multiple self-hosted runners active. It is possible that different runners will pickup different jobs on a given GitHub Actions run. Therefore, the folder mounted on each job may be different, leading to errors. Bypassing the default `--workdir` to a folder that will be accessible by all self-hosted runners would solve this problem. It would be great if developers provide a fix for this.

**Comment by paige-wdc** [comment]
This is also something that my team would find very useful.

**Comment by reesericci** [comment]
This is just ridiculous - this bug has been around for two years, and nothing has been done about it. GitHub really cares about it's "community".

**Comment by prashantkalkar** [comment]
I am working around this limitation by providing a default workspace for the job. As follows:

```
jobs:
  <job_name>:
     container:
        image: ubuntu:20.04
     defaults:
        run:
           working-directory: /
```
Now any steps in the job will run as per new working directory. May be this is useful till the actual solution is implemented.

**Comment by neilime** [comment]
This is issue is really painfull, because not all actions manage custom working directory, plus if current working directory is specific (Example: `/var/www/html`), github actions is mounting action container on `/github/workflow` in this cas it is impossible to reach the files, unless doing workaround with symlinks...

**Comment by dimamovic-exos** [comment]
> I am working around this limitation by providing a default workspace for the job. As follows:
> 
> ```
> jobs:
>   <job_name>:
>      container:
>         image: ubuntu:20.04
>      defaults:
>         run:
>            working-directory: /
> ```
> 
> Now any steps in the job will run as per new working directory. May be this is useful till the actual solution is implemented.
> 
> Note that, the marketplace actions might still work with the GITHUB_WORKSPACE location. So configure the action appropriately.

This doesn't work for me. Did you try with diff working directory something like `working-directory: /dir1/dir2` ?

**Comment by shqear93** [comment]
> This is issue is really painfull, because not all actions manage custom working directory, plus if current working directory is specific (Example: `/var/www/html`), github actions is mounting action container on `/github/workflow` in this cas it is impossible to reach the files.
> 
> Here is the workaround I found but not proud of it:
> 
> ```yaml
> jobs:
>   test:
>     runs-on: "ubuntu-latest"
>       image: my-image:latest
>       options:
>         --user 1001:1001
>         -v "${{github.workspace}}":"${{github.workspace}}"
>     steps:
>       - run: cp -R /var/www/html/* ${{github.workspace}}/
> 
>       - uses: actions/sample-action@main
> ```

It results with error
```
The workflow is not valid. .github/workflows/build-and-deploy.yml (Line: 72, Col: 7): Unexpected value 'defaults'
```

in the end, I updated my commands to "cd /app" first

**Comment by shqear93** [comment]
> I am working around this limitation by providing a default workspace for the job. As follows:
> 
> ```
> jobs:
>   <job_name>:
>      container:
>         image: ubuntu:20.04
>      defaults:
>         run:
>            working-directory: /
> ```
> 
> Now any steps in the job will run as per new working directory. May be this is useful till the actual solution is implemented.
> 
> Note that, the marketplace actions might still work with the GITHUB_WORKSPACE location. So configure the action appropriately.

It results with error
```
The workflow is not valid. .github/workflows/build-and-deploy.yml (Line: 72, Col: 7): Unexpected value 'defaults'
```

in the end, I updated my commands to "cd /app" first

**Comment by Atomicchipmonk** [comment]
upvote - this would really help with self-hosted runner isolation

**Comment by vt-rcheng** [comment]
yes, the container job should also not force mounting of volumes

`-v "/home/<user>/actions-runner/_work":"/__w" -v "/home/<user>/actions-runner/externals":"/__e":ro -v "/home/<user>/actions-runner/_work/_temp":"/__w/_temp" -v "/home/<user>/actions-runner/_work/_actions":"/__w/_actions" -v "/home/<user>/actions-runner/_work/_tool":"/__w/_tool" -v "/home/<user>/actions-runner/_work/_temp/_github_home":"/github/home" -v "/home/<user>/actions-runner/_work/_temp/_github_workflow":"/github/workflow"`

please give control back to the developer.
if I want to mount these volumes I would do it myself or make it an optional config setting.
